Add HMAC signature and adapt to Python 3

eliemichel opened the pull request:
### Add signature to zmq messages (HMAC digest)

So we can remove option `--Session.key=''` in call to ipython
And so we fix #41 since it then supports Python 3

/!\ Signature of `util.ipyEncodeAndSend` changed:
I added the session key as thrid argument.

/!\ New dependency: luacrypto
Where do I have to notify it?

/!\ `d:final()` used instead of `d:digest()`
According to the [documentation](http://luacrypto.luaforge.net/manual.html#reference), `hmac` object should have a `digest` method, but it is in fact named `final`.
### Add some missing spaces
### Fix formatting issue in `CONTRIBUTING.md`
### Add Arch Linux to the list of tested systems

_NB: I read the guidelines for contributing, but:_
1. I could not run test.lua in its own. I do I run tests?
2. I changed the API of a fonction, but there is actually no documentation to change.
3. The code I patch use 3 space indentation, so I respected it despite the guidelines.
4. Same for the 80 characters constraint.

eliemichel committed:
Add HMAC signature and adapt to Python 3

Add signature to zmq messages (HMAC digest)
So we can remove option --Session.key='' in call to ipython
And so we fix #41 and then support Python 3

Add some missing spaces (coding style)

/!\ Signature of `util.ipyEncodeAndSend` changed:
I added the session key as thrid argument.

/!\ New dependency: luacrypto

/!\ `d:final()` used instead of `d:digest()`
According to the documentation
(http://luacrypto.luaforge.net/manual.html#reference)
hmac object should have a `digest` method, but it is in fact named `final`

eliemichel committed:
Fix formatting issue in CONTRIBUTING.md

eliemichel committed:
Add Arch Linux to the tested distros

soumith commented:
@eliemichel You missed changing quite a few calls of util.ipyEncodeAndSend, in IOHandler.lua, Plot.lua, gfx.lua and stuff. IOHandler.lua is especially critical, because it's the process that forward all the stdout/stderr to ipython

eliemichel commented:
Oops I completely forgot it… And I will have to propagate this argument to some other fonctions.

soumith commented:
@eliemichel sure, setting util.key sounds okay to me.

However, on part you have to change is how you communicate the key to IOHandler.
I already send a couple of parameters from main to IOHandler over with something called private_msg, you could probably do the same:

https://github.com/facebook/iTorch/blob/master/main.lua#L150
https://github.com/facebook/iTorch/blob/master/IOHandler.lua#L91

eliemichel commented:
@soumith `IOHandler` can access `ipycfg` too, actually.

soumith commented:
oh you're right, good point. Sure, as a setter sounds fine too. Feel free to do your take on this, I am not particular about how key gets passed to util

eliemichel committed:
Add session_key attribute to util module

It is not required as a ipyEncodeAndSend argument any more, thus we do not have
to find a way to propagate it.

`session_key` is private and is set through `setSessionKey`.

No `getSessionKey` has been created yet as we have no use of it, but it can be
easily added.

eliemichel commented:
@soumith updated

eliemichel commented:
@soumith Is there something wrong with this PR?

I have had no problem with it during the last week.

soumith commented:
hey @eliemichel nothing wrong. I am just testing it out on multiple platforms before I pull it in. Sorry for the delay.

eliemichel commented:
No problem! I was just lacking feedback but I understand well that you want to stress-test it before any merging.

soumith commented:
Thanks @eliemichel ! I've merged it. 
There seems to be some small issue, finding OpenSSL on OSX is not easy for luacrypto. I am seeing how this can be worked around.

eliemichel commented:
Well, I just called `luarocks install luacrypto`. But indeed I do not use OSX. (I tested Arch Linux and Ubuntu).

Actually I guess there is a way to add this dependency so that `luarocks make` would install it, but I do not know it.
